File: packages/php-sdk/src/Webhooks/Verifier.php
<?php

declare(strict_types=1);

namespace PayXCommerce\Webhooks;

use PayXCommerce\Exceptions\WebhookVerificationException;

final class Verifier
{
    public function __construct(
        private readonly string $webhookSecret,
        private readonly int $toleranceSeconds = 300,
    ) {
    }

    public function verify(string $rawBody, array $headers): array
    {
        $eventId = $this->header($headers, 'X-PXC-Event-ID');
        $timestamp = $this->header($headers, 'X-PXC-Timestamp');
        $signature = $this->header($headers, 'X-PXC-Signature');

        if (!$eventId || !$timestamp || !$signature) {
            throw new WebhookVerificationException('Missing PayXCommerce webhook signature headers.');
        }

        if (!ctype_digit((string) $timestamp)) {
            throw new WebhookVerificationException('Invalid PayXCommerce webhook timestamp.');
        }

        if (abs(time() - (int) $timestamp) > $this->toleranceSeconds) {
            throw new WebhookVerificationException('PayXCommerce webhook timestamp is outside the allowed tolerance.');
        }

        $payload = json_decode($rawBody, true);
        if (!is_array($payload)) {
            throw new WebhookVerificationException('PayXCommerce webhook body is not valid JSON.');
        }

        $expected = self::signature($eventId, $rawBody, $this->webhookSecret);
        if (!hash_equals($expected, $signature)) {
            throw new WebhookVerificationException('Invalid PayXCommerce webhook signature.');
        }

        return $payload;
    }

    public static function signature(string $eventId, string $rawBody, string $webhookSecret): string
    {
        $payload = json_decode($rawBody, true);
        $canonicalBody = is_array($payload) ? json_encode($payload, JSON_UNESCAPED_SLASHES) : $rawBody;

        return hash_hmac('sha256', $eventId . '.' . $canonicalBody, $webhookSecret);
    }

    private function header(array $headers, string $name): ?string
    {
        foreach ($headers as $key => $value) {
            if (strcasecmp((string) $key, $name) === 0) {
                return is_array($value) ? (string) ($value[0] ?? '') : (string) $value;
            }
        }

        $serverName = 'HTTP_' . strtoupper(str_replace('-', '_', $name));
        if (isset($headers[$serverName])) {
            return (string) $headers[$serverName];
        }

        return null;
    }
}

File: packages/python-sdk/payxcommerce/webhooks/verifier.py
from __future__ import annotations

import hashlib
import hmac
import json
import time
from typing import Any

from ..exceptions import WebhookVerificationException


class Verifier:
    def __init__(self, webhook_secret: str, tolerance_seconds: int = 300):
        self.webhook_secret = webhook_secret
        self.tolerance_seconds = tolerance_seconds

    def verify(self, raw_body: str, headers: dict[str, str]) -> dict[str, Any]:
        event_id = self._header(headers, "X-PXC-Event-ID")
        timestamp = self._header(headers, "X-PXC-Timestamp")
        signature = self._header(headers, "X-PXC-Signature")
        if not event_id or not timestamp or not signature:
            raise WebhookVerificationException("Missing PayXCommerce webhook signature headers.")
        if not timestamp.isdigit():
            raise WebhookVerificationException("Invalid PayXCommerce webhook timestamp.")
        if abs(int(time.time()) - int(timestamp)) > self.tolerance_seconds:
            raise WebhookVerificationException("PayXCommerce webhook timestamp is outside the allowed tolerance.")
        try:
            payload = json.loads(raw_body)
        except json.JSONDecodeError as exc:
            raise WebhookVerificationException("PayXCommerce webhook body is not valid JSON.") from exc
        expected = self.signature(event_id, raw_body, self.webhook_secret)
        if not hmac.compare_digest(expected, signature):
            raise WebhookVerificationException("Invalid PayXCommerce webhook signature.")
        return payload

    @staticmethod
    def signature(event_id: str, raw_body: str, webhook_secret: str) -> str:
        try:
            payload = json.loads(raw_body)
            canonical_body = json.dumps(payload, separators=(",", ":"), ensure_ascii=False)
        except json.JSONDecodeError:
            canonical_body = raw_body
        return hmac.new(webhook_secret.encode(), f"{event_id}.{canonical_body}".encode(), hashlib.sha256).hexdigest()

    def _header(self, headers: dict[str, str], name: str) -> str:
        normalized = name.lower()
        for key, value in headers.items():
            if key.lower() == normalized:
                return str(value)
        server_name = "HTTP_" + name.upper().replace("-", "_")
        return str(headers.get(server_name, ""))
